### ISSUE TITLE
Tracking Issue for RFC 3216: "Allow using `for<'a>` syntax when declaring closures"

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#97362]
This is a tracking issue for the RFC "Allow using `for<'a>` syntax when declaring closures" (rust-lang/rfcs#3216).
The feature gate for the issue is `#![feature(closure_lifetime_binder`)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [ ] Implement the RFC (cc @rust-lang/XXX -- can anyone write up mentoring
      instructions?)
- [ ] Adjust documentation ([see instructions on rustc-dev-guide][doc-guide])
- [ ] Stabilization PR ([see instructions on rustc-dev-guide][stabilization-guide])

[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr
[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs

### Unresolved Questions

None at this time

**Comment by WaffleLapkin** [comment]
I'd like to work on implementing this 👀 

@rustbot claim

**Comment by fakedrake** [comment]
I bumped on this issue and I would like to share some code that maybe someone knowledgeable could comment on whether we are likely to get the features demonstrated or similar (in summary it's getting `for<>` for async closures):

```rust
#![feature(closure_lifetime_binder)]
#![feature(trait_alias)]

use core::future::Ready;
use core::future::ready;
use core::future::Future;

// Don't read too much into the implementation of `S`
struct S<'a>(&'a ());

trait AsyncF<'a, F> = Fn(S<'a>) -> F where F: Future<Output = S<'a>>;
async fn asyncF<F>(f: impl for<'a> AsyncF<'a, F>) {
    let a = ();
    f(S(&a)).await;
}

fn main() {
    // We can't seem to do this because for<> does not seem to extend to the
    // argument.
    asyncF(for<'a> |x: S<'a>| -> Ready<S<'a>> {ready(x)});

    
    // Ideally we would like this to do anything useful with for<> and async.
    // asyncF(for<'a> async move |x: &'a ()| -> &'a () {ready(x)});
}
```
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=75475a141e376a2c0711ecac8461d20e)

**Comment by WaffleLapkin** [comment]
@fakedrake I think your `AsyncF`/`asyncF` are wrong. You are using a single `F` type, that is supposed to implement `for<'a> Future<Output = S<'a>>`, where in reality you have a `for<'a> F<'a>: Future<Output = S<'a>>` type.

If you change the `AsyncF` trait to be something like the following:
```rust
trait AsyncF<'a, Input>: Fn(Input) -> Self::Future {
    type Future: Future<Output = Self::Out>;
    
    type Out;
}

impl<'a, F, Fut, In> AsyncF<'a, In> for F
where
    F: Fn(In) -> Fut,
    Fut: Future,
{
    type Future = Fut;
    type Out = Fut::Output;
}
```
Then your example [works](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=a0db8d120af26cbcab4e20dc309b4adc). Though I want to warn you that such traits are a pain to work with and especially extend.

----

As for `for<'a> async move |x: &'a ()| -> &'a ()` kind of thing, I think it generally desired, but was ruled out of the scope of the RFC. So, if we'll ever add it, it would first need its own RFC.

It appears that async closures are currently in a very rough shape ([eg](https://github.com/rust-lang/rust/blob/b8c0a01b2b651416f5e7461209ff1a93a98619e4/compiler/rustc_ast_lowering/src/expr.rs#L962)) so a good place to start would probably be to fix/complete the implementation of async closures first.

**Comment by fakedrake** [comment]
Thank you very much. I have seen people vaguely complaining about the primitive state of async closures but besides this niche case I have found that they have worked pretty well. Is there an issue tracking the problems with them?

**Comment by WaffleLapkin** [comment]
There is a tracking issue for async closures: #62290

**Comment by Jules-Bertholet** [comment]
Found an ICE involving this feature: https://github.com/rust-lang/rust/issues/103736

(Maybe we could use an `F-closure_lifetime_binder` tag?)

**Comment by Jules-Bertholet** [comment]
Suggestion: could it be allowed to elide the return type of an explicitly higher-ranked closure when it is `()`? Currently, Rust allows eliding `()` returns in all other function signatures, even when nothing else can be elided.

**Comment by WaffleLapkin** [comment]
@Jules-Bertholet I don't think it's a good idea, we may want to allow inferring return type in the future, just like with normal closures.

**Comment by Jules-Bertholet** [comment]
@WaffleLapkin "Infer return type when it is `()`" should be forward-compatible with "infer return type always."

**Comment by WaffleLapkin** [comment]
@Jules-Bertholet oh, I misunderstood this as "make `for<> || { some expression... }` return type always `()`, same as with normal function". Inferring it is probably fine.

**Comment by kevincox** [comment]
It appears that this will not work well for returning closures as their type can not be explicitly specified as a closure return type. This is likely known but I think worth pointing it out. This will however work for return types that you can specify.

```rust
let f = for<'a> |a: &'a u8| -> impl 'a + Fn() -> &'a u8 {
    || a
};
```

```
error[[E0562]](https://doc.rust-lang.org/nightly/error_codes/E0562.html): `impl Trait` only allowed in function and inherent method return types, not in closure return types
 --> src/main.rs:5:36
  |
5 |     let f = for<'a> |a: &'a u8| -> impl 'a + Fn() -> &'a u8 {
  |                                    ^^^^^^^^^^^^^^^^^^^^^^^^
```

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7aa83bd601924e751507bc7062715256)

**Comment by fee1-dead** [comment]
@kevincox: I believe that is a general limitation of what return types you can write when declaring closures, and not related to this RFC.

**Comment by zirconium-n** [comment]
Is this feature blocked by any ongoing process? The implementation seems completed for a while, maybe we can push this to stabilization?

**Comment by opsnull** [comment]
> Is this feature blocked by any ongoing process? The implementation seems completed for a while, maybe we can push this to stabilization?

same question here.

**Comment by runiq** [comment]
A question out of interest: Is this in any way a prerequisite for https://github.com/rust-lang/rust/issues/68923?

**Comment by safinaskar** [comment]
Unfortunately, `for<'a> |x: &'a ()| -> &'a () { x }` and `|x: &()| -> &() { x }` currently are not equivalent. Here is minimal example:

```rust
#![feature(closure_lifetime_binder)]
#![feature(unboxed_closures)]

fn foo<F>(f: F)
where
    F: for<'a> FnOnce<(&'a (),)>,
{}

fn main() {
    foo(for<'a> |x: &'a ()| -> &'a () { x }); // COMPILES
    foo(|x: &()| -> &() { x }); // DOESN'T COMPILE
}
```

Here is playground:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=47b1d8af93240be0cc1326a892dc01fd

This was discovered because of https://github.com/rust-lang/rust/issues/34162 . Here is (larger) code, which tries to fix `sort_by_key` (by [Jules-Bertholet](https://github.com/Jules-Bertholet)):

```rust
#![feature(closure_lifetime_binder)]
#![feature(unboxed_closures)]

fn sort_by_key<T, F>(s: &mut [T], mut f: F)
where
    F: for<'a> FnMut<(&'a T,)>,
    // instead of `B: Ord`
    for<'a> <F as FnOnce<(&'a T,)>>::Output: Ord,
{
    s.sort_by(|a, b| f(a).cmp(&f(b)))
}

#[derive(Debug)]
struct Client(String);

impl Client {
    fn key(&self) -> &str {
        &self.0
    }
}

fn main() {
    let mut test = vec![
        Client("c".to_string()),
        Client("a".to_string()),
        Client("b".to_string()),
    ];
    sort_by_key(&mut test, for<'a> |c: &'a Client| -> &'a str { c.key() }); // compiles
    //sort_by_key(&mut test, |c: &Client| -> &str { c.key() }); // doesn't compile
    dbg!(test);
}
```

The code above compiles on nightly without any additional flags (assuming you choose `for<'a>` version)

Also, here is another problematic code, which doesn't require `unboxed_closures`:
```rust
#![feature(closure_lifetime_binder)]

trait X<'a> {
}

impl<'a, T, K> X<'a> for T
where
    T: FnOnce(&'a ()) -> K
{}

fn foo<F>(f: F)
where
    F: for<'a> X<'a>,
{}

fn main() {
    foo(for<'a> |x: &'a ()| -> &'a () { x }); // COMPILES
    foo(|x: &()| -> &() { x }); // DOESN'T COMPILE
}
```

playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=bfcdede24f2e7b27380e2990df1af328

**Comment by workingjubilee** [comment]
So you're saying we need the syntactic ability to specifically annotate all closure lifetimes in order to be able to write the programs we want to write? Some kind of `for<'lt>` syntax, maybe?

**Comment by safinaskar** [comment]
@workingjubilee , no. We already have such syntax, it is available with `#![feature(closure_lifetime_binder)]`. I'm just saying `for<'a> |x: &'a ()| -> &'a () { x }` should work exactly like `|x: &()| -> &() { x }`, but it is currently not. And this causes problems as I showed above, which prevent us from having proper ergonomic `sort_by_key`

**Comment by Jules-Bertholet** [comment]
*Ideally*, the compiler would infer the most permissive signature possible, and so the `for<'a>` syntax would not be necessary (except as documentation).

**Comment by workingjubilee** [comment]
>  I'm just saying `for<'a> |x: &'a ()| -> &'a () { x }` should work exactly like `|x: &()| -> &() { x }`, but it is currently not.

It is my understanding that the binding of a specific lifetime and a `for<'lt>` (which, as I understand it, is effectively the ∀ lifetime) are not interchangeable, for the precise reason that ∀ and ∃ are different quantifiers.

**Comment by WanderLanz** [comment]
@safinaskar As far as I understand compiler internals, a closure is inherently not invariant in regards to the lifetime of the arguments, and you need strong hints in order to make a closure such similar to a normal `fn`. For example, the higher ranked closure macros in the `lending_iterator` crate. 

In other words, the compiler is very conservative when closures could possibly have something like a `for<'all> where Closure: 'all` bound.

```rust
#![feature(closure_lifetime_binder)]

trait X<'a> {
}

impl<'a, T, K> X<'a> for T
where
    T: FnOnce(&'a ()) -> K
{}

fn foo<F>(f: F)
where
    F: for<'a> X<'a>,
{}

fn main() {
    foo(for<'a> |x: &'a ()| -> &'a () { x }); // COMPILES
    // foo(|x: &()| -> &() { x }); // DOESN'T COMPILE
    
    fn hrc_hint<F>(f: F) -> F
    where
        F: for<'all> FnOnce(&'all ()) -> &'all ()
    {
        f
    }
    foo(hrc_hint(|x: &()| -> &() { x })); // NOW COMPILES
}
```

Finally, to avoid this, as far as I can tell, either: the compiler becomes less conservative on closures (possibly E.2024?), we get bounded HRTBs (`for<'all> where *: 'all`) (very unlikely), or `closure_lifetime_binder` becomes popular enough to be stabilized.

**Comment by Kixunil** [comment]
@runiq I don't know if it's required, it certainly looks helpful but FYI it's not *suffficient* - I tried using this and the code still doesn't compile.

**Comment by vojtechkral** [comment]
Hi!
I thought the code from #129787 could be fixed with this feature, but it looks like I'm getting the same error still:

```rust
fn f() -> impl Future<Output = ()> + Send {
    async {
        let data: &[_] = &[0u8];
        stream::iter([data])
            .map(for<'x> |_data: &'x [u8]| -> future::Ready<()> { future::ready(()) })
            .buffer_unordered(1)
            .for_each_concurrent(4, |_| async {})
            .await;
    }
}
```

-> ``implementation of `FnOnce` is not general enough``.

though when I pass the lambda through an identity function that explicitly declares/requires a HRTB lifetime, it then builds:

```rust
fn lifetime_hrtb<F, R>(f: F) -> F
where
    F: for<'a> FnMut(&'a [u8]) -> R,
{
    f
}

fn f() -> impl Future<Output = ()> + Send {
    async {
        let data: &[_] = &[0u8];
        stream::iter([data])
            .map(lifetime_hrtb(|_data| async {}))
            .buffer_unordered(1)
            .for_each_concurrent(4, |_| async {})
            .await;
    }
}
```

any clues as to what's going on? 🙏

**Comment by compiler-errors** [comment]
This is a tracking issue, for the record. If you have a bug with a specific feature, please open a new issue!
